### DISCUSSION TITLE
Middleware是否可以支持网络请求？

### DISCUSSION THREAD

**Discussion by danatechgithub4** [junerver/ComposeHooks#12] [Q&A] [answered]
貌似Middleware的实现过于简单跟Redux Thunk不太一样，
不太确定，我实现一个类似logMiddleware的Middleware，里面执行网络等耗时请求，会不会有问题？

或者我可以用useDispatchAsync。

希望作者结合一下useRequest和TodosListContainer()完善example具体使用方式，我不太确定目前用的对不对！

**Comment by junerver** [comment]
不可以，虽然叫中间件，但是和 redux-js 还是有很大的区别，起初我只是为了实现无侵入的log，至于类似 redux-thunk这样的异步操作，其实是通过 `useDispatchAsync` 实现的。如果你需要执行网络请求，你应该使用`useDispatchAsync`，或者可以参考它的实现再进行自己的封装；

目前中间件必须要是一个纯函数，否则他会破坏 useReducer；

至于示例example，我没理解你有什么具体的疑惑？或者你可以在github上创建一个demo项目，然后我们在这个项目里具体分析一下？

**Reply by danatechgithub4** [reply]
![image](https://github.com/junerver/ComposeHooks/assets/127659297/8a9d324e-3dbe-461a-b19f-370e401fcebc)

想要一个useRequest结合useDispatchAsync的demo。

网络请求成功和失败都要考虑进来。

**Comment by junerver** [comment]
> ![image](https://private-user-images.githubusercontent.com/127659297/330692864-8a9d324e-3dbe-461a-b19f-370e401fcebc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3NTk0MjAsIm5iZiI6MTcxNTc1OTEyMCwicGF0aCI6Ii8xMjc2NTkyOTcvMzMwNjkyODY0LThhOWQzMjRlLTNkYmUtNDYxYS1iMTlmLTM3MGU0MDFmY2ViYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxNVQwNzQ1MjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYWExNjU2ZGM3MzU0MjUyYmUwNDczMWJmYzExZGM0MzljZGUxNzc3M2E1MmVjZDhmMmU5Y2Y1YjExZGE1YTU0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.xnnHFbgeybH2NiqlJMxu6_pCGOvEogP3N2A9XtOu-go)
> 
> 想要一个useRequest结合useDispatchAsync的demo。
> 
> 网络请求成功和失败都要考虑进来。

下面的 `UseReduxFetch2` 这个不能满足需求么？其中的 `useFetch` 这个自定义封装就是考虑了成功失败处理的

**Comment by junerver** [comment]
另外 `useRequest` 已经是一个功能完毕的封装了，如果只是用来网络请求为什么不直接使用 `useRequest` 呢？

**Reply by danatechgithub4** [reply]
拿`TodosListContainer()`例子来说，我删除一个Todo，调用网络接口成功后，网络接口只会返回给成功或者失败，成功后，我再调用`dispatch(DelTodo(item.id))`。直接`useRequest`怎么做？

失败后弹个弹窗提示一下。

**Comment by junerver** [comment, marked as answer]
> 拿`TodosListContainer()`例子来说，我删除一个Todo，调用网络接口成功后，网络接口只会返回给成功或者失败，成功后，我再调用`dispatch(DelTodo(item.id))`。直接`useRequest`怎么做？
> 
> 失败后弹个弹窗提示一下。

你想表达的是不是：
- 接口1 ：列表接口
- 接口2：del接口

期望：
1. 进入页面请求`接口1`，获取列表渲染
2. 点击del、发起请求`接口2`
3. 根据 步骤2，提示 or 更新列表

这个需求完全可以通过 `useRequest` 达成
将这两个接口分别使用useRequest创建，然后解构出接口1的状态、refresh函数；
接口2设置为手动出发，解构出 run 函数，同时注册生命周期函数：onSuccess 时调用1的 refresh、onError 时回调 toast

**Reply by danatechgithub4** [reply]
不需要接口1 ：列表接口，这个列表已经在启动页缓存在内存里面了（做了持久化）。
伪代码如下：
![image](https://github.com/junerver/ComposeHooks/assets/127659297/afaac346-da8f-49f1-8835-06ff501da5c5)

**Comment by junerver** [comment]
你可以参考：
- [manual and automatic](https://github.com/junerver/ComposeHooks/blob/master/app/src/main/java/xyz/junerver/composehooks/example/request/Auto%26Manual.kt)
- [life cycle callbacks](https://github.com/junerver/ComposeHooks/blob/master/app/src/main/java/xyz/junerver/composehooks/example/request/Lifecycle.kt)
- [refresh](https://github.com/junerver/ComposeHooks/blob/master/app/src/main/java/xyz/junerver/composehooks/example/request/Refresh.kt)

**Comment by junerver** [comment]
> 不需要接口1 ：列表接口，这个列表已经在启动页缓存在内存里面了（做了持久化）。 伪代码如下： ![image](https://private-user-images.githubusercontent.com/127659297/330719009-afaac346-da8f-49f1-8835-06ff501da5c5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3NjMzMzYsIm5iZiI6MTcxNTc2MzAzNiwicGF0aCI6Ii8xMjc2NTkyOTcvMzMwNzE5MDA5LWFmYWFjMzQ2LWRhOGYtNDlmMS04ODM1LTA2ZmY1MDFkYTVjNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxNVQwODUwMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NGJiMDY5MWYxNTJmMGVmZTAxMjRkYTRiMTJiN2ZkOGU0YWVlMGRkZDFhMDRhZTc0M2MyNmUyNmM4MDc2OTk4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bETMWmzNwsIhUXjIf-Ls0YBXhwl43s0USG_NrKtDcY4)

那就更简单了，你只需要使用生命周期回调即可
